### ISSUE TITLE
Uncaught Error: Class 'MZ_Mobilize_America\ShortCode\ShortCode_Script_Loader' not found

### ISSUE THREAD

**Issue by jbradach** [MikeiLL/mz-mobilize-america-interface#3]
Thanks for your work on this plugin. It is exactly what I've been looking for now that Mobilize doesn't recommend directly embedding their feeds on sites.

When I tried activating this plugin, I encountered a Composer-related message. `composer install` was sufficient to resolve this.

```bash
mobilize@RZm-mobilize:~/public/wp-content/plugins/mobilize-america$ wp plugin activate mobilize-america
Mobilize America requires Composer autoloading, which is not configured
```

Attempting to activate the plugin a second time results in a class not found error.

```php
PHP Fatal error:  Uncaught Error: Class 'MZ_Mobilize_America\ShortCode\ShortCode_Script_Loader' not found in /www/mobilize_211/public/wp-content/plugins/mobilize-america/src/Display/Display.php:9
Stack trace:
#0 phar:///usr/local/bin/wp/vendor/composer/ClassLoader.php(444): include()
#1 phar:///usr/local/bin/wp/vendor/composer/ClassLoader.php(322): Composer\Autoload\includeFile('/www/mobilize_2...')
#2 [internal function]: Composer\Autoload\ClassLoader->loadClass('MZ_Mobilize_Ame...')
#3 /www/mobilize_211/public/wp-content/plugins/mobilize-america/src/Core/Plugin_Core.php(142): spl_autoload_call('MZ_Mobilize_Ame...')
#4 /www/mobilize_211/public/wp-content/plugins/mobilize-america/src/Core/Plugin_Core.php(70): MZ_Mobilize_America\Core\Plugin_Core->define_public_hooks()
#5 /www/mobilize_211/public/wp-content/plugins/mobilize-america/mobilize-america.php(100): MZ_Mobilize_America\Core\Plugin_Core->__construct()
#6 /www/mobilize_211/public/wp-content/plugins/mobilize-america/mobilize-america.php(120): MZ_Mobilize_America\Mobilize_ in /www/mobilize_211/public/wp-content/plugins/mobilize-america/src/Display/Display.php on line 9
Fatal error: Uncaught Error: Class 'MZ_Mobilize_America\ShortCode\ShortCode_Script_Loader' not found in /www/mobilize_211/public/wp-content/plugins/mobilize-america/src/Display/Display.php:9
Stack trace:
#0 phar:///usr/local/bin/wp/vendor/composer/ClassLoader.php(444): include()
#1 phar:///usr/local/bin/wp/vendor/composer/ClassLoader.php(322): Composer\Autoload\includeFile('/www/mobilize_2...')
#2 [internal function]: Composer\Autoload\ClassLoader->loadClass('MZ_Mobilize_Ame...')
#3 /www/mobilize_211/public/wp-content/plugins/mobilize-america/src/Core/Plugin_Core.php(142): spl_autoload_call('MZ_Mobilize_Ame...')
#4 /www/mobilize_211/public/wp-content/plugins/mobilize-america/src/Core/Plugin_Core.php(70): MZ_Mobilize_America\Core\Plugin_Core->define_public_hooks()
#5 /www/mobilize_211/public/wp-content/plugins/mobilize-america/mobilize-america.php(100): MZ_Mobilize_America\Core\Plugin_Core->__construct()
#6 /www/mobilize_211/public/wp-content/plugins/mobilize-america/mobilize-america.php(120): MZ_Mobilize_America\Mobilize_ in /www/mobilize_211/public/wp-content/plugins/mobilize-america/src/Display/Display.php on line
```

I'm using PHP 7.3-FPM with Nginx 1.19.1 on Ubuntu 20.04. Details below.

```shell
mobilize@RZm-mobilize:~$ uname -a
Linux RZm-mobilize 5.3.0-1018-gcp #19~18.04.1-Ubuntu SMP Tue Apr 14 12:49:45 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

mobilize@RZm-mobilize:~$ php -v
PHP 7.3.20-1+ubuntu20.04.1+deb.sury.org+1 (cli) (built: Jul  9 2020 16:34:52) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.3.20, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.3.20-1+ubuntu20.04.1+deb.sury.org+1, Copyright (c) 1999-2018, by Zend Technologies
[28-Jul-2020 02:53:59] ERROR: FPM initialization failed

mobilize@RZm-mobilize:~/public/wp-content/plugins/mobilize-america$ php-fpm7.3 -v
PHP 7.3.20-1+ubuntu20.04.1+deb.sury.org+1 (fpm-fcgi) (built: Jul  9 2020 16:34:52)
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.3.20, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.3.20-1+ubuntu20.04.1+deb.sury.org+1, Copyright (c) 1999-2018, by Zend Technologies

mobilize@RZm-mobilize:~/public/wp-content/plugins/mobilize-america$ nginx -V
nginx version: nginx/1.19.1
built with OpenSSL 1.1.1f  31 Mar 2020 (running with OpenSSL 1.1.1g  21 Apr 2020)
TLS SNI support enabled
configure arguments: --with-openssl-opt=enable-tls1_3 --with-cc-opt='-g -O2 -fdebug-prefix-map=/build/nginx-gN4tCp/nginx-1.19.1=. -fstack-protector-strong -Wformat -Werror=format-security -fPIC -Wdate-time -D_FORTIFY_SOURCE=2' --with-ld-opt='-Wl,-Bsymbolic-functions -Wl,-z,relro -Wl,-z,now -fPIC' --prefix=/usr/share/nginx --conf-path=/etc/nginx/nginx.conf --http-log-path=/var/log/nginx/access.log --error-log-path=/var/log/nginx/error.log --lock-path=/var/lock/nginx.lock --pid-path=/run/nginx.pid --modules-path=/usr/lib/nginx/modules --http-client-body-temp-path=/var/lib/nginx/body --http-fastcgi-temp-path=/var/lib/nginx/fastcgi --http-proxy-temp-path=/var/lib/nginx/proxy --http-scgi-temp-path=/var/lib/nginx/scgi --http-uwsgi-temp-path=/var/lib/nginx/uwsgi --with-debug --with-pcre-jit --with-http_ssl_module --with-http_stub_status_module --with-http_realip_module --with-http_auth_request_module --with-http_v2_module --with-http_dav_module --with-http_slice_module --with-threads --with-http_addition_module --with-http_flv_module --with-http_geoip_module=dynamic --with-http_gunzip_module --with-http_gzip_static_module --with-http_image_filter_module=dynamic --with-http_mp4_module --with-http_perl_module=dynamic --with-http_random_index_module --with-http_secure_link_module --with-http_sub_module --with-http_xslt_module=dynamic --with-mail=dynamic --with-mail_ssl_module --with-stream=dynamic --with-stream_ssl_module --with-stream_ssl_preread_module --add-dynamic-module=/build/nginx-gN4tCp/nginx-1.19.1/debian/modules/http-headers-more-filter --add-dynamic-module=/build/nginx-gN4tCp/nginx-1.19.1/debian/modules/http-auth-ldap --add-dynamic-module=/build/nginx-gN4tCp/nginx-1.19.1/debian/modules/http-auth-pam --add-dynamic-module=/build/nginx-gN4tCp/nginx-1.19.1/debian/modules/http-cache-purge --add-dynamic-module=/build/nginx-gN4tCp/nginx-1.19.1/debian/modules/http-dav-ext --add-dynamic-module=/build/nginx-gN4tCp/nginx-1.19.1/debian/modules/http-ndk --add-dynamic-module=/build/nginx-gN4tCp/nginx-1.19.1/debian/modules/http-echo --add-dynamic-module=/build/nginx-gN4tCp/nginx-1.19.1/debian/modules/http-redis2 --add-dynamic-module=/build/nginx-gN4tCp/nginx-1.19.1/debian/modules/http-fancyindex --add-dynamic-module=/build/nginx-gN4tCp/nginx-1.19.1/debian/modules/http-brotli --add-dynamic-module=/build/nginx-gN4tCp/nginx-1.19.1/debian/modules/nchan --add-dynamic-module=/build/nginx-gN4tCp/nginx-1.19.1/debian/modules/njs/nginx --add-dynamic-module=/build/nginx-gN4tCp/nginx-1.19.1/debian/modules/http-lua --add-dynamic-module=/build/nginx-gN4tCp/nginx-1.19.1/debian/modules/rtmp --add-dynamic-module=/build/nginx-gN4tCp/nginx-1.19.1/debian/modules/http-uploadprogress --add-dynamic-module=/build/nginx-gN4tCp/nginx-1.19.1/debian/modules/http-upstream-fair --add-dynamic-module=/build/nginx-gN4tCp/nginx-1.19.1/debian/modules/http-subs-filter --add-dynamic-module=/build/nginx-gN4tCp/nginx-1.19.1/debian/modules/http-vhost-traffic-status

mobilize@RZm-mobilize:~/public/wp-content/plugins/mobilize-america$ wp core version --extra
WordPress version: 5.4.2
Database revision: 47018
TinyMCE version:   4.960 (4960-20190918)
Package language:  en_US
```

**Comment by MikeiLL** [comment]
Sorry for the trouble. 

UPDATE see below

Do you have the following file:

`public/wp-content/plugins/mobilize-america/vendor/composer/autoload_psr4.php` with following content:

```
<?php

// autoload_psr4.php @generated by Composer

$vendorDir = dirname(dirname(__FILE__));
$baseDir = dirname($vendorDir);

return array(
    'Symfony\\Polyfill\\Php80\\' => array($vendorDir . '/symfony/polyfill-php80'),
    'Symfony\\Polyfill\\Php73\\' => array($vendorDir . '/symfony/polyfill-php73'),
    'Symfony\\Polyfill\\Mbstring\\' => array($vendorDir . '/symfony/polyfill-mbstring'),
    'Symfony\\Polyfill\\Intl\\Normalizer\\' => array($vendorDir . '/symfony/polyfill-intl-normalizer'),
    'Symfony\\Polyfill\\Intl\\Grapheme\\' => array($vendorDir . '/symfony/polyfill-intl-grapheme'),
    'Symfony\\Polyfill\\Ctype\\' => array($vendorDir . '/symfony/polyfill-ctype'),
    'Symfony\\Contracts\\Service\\' => array($vendorDir . '/symfony/service-contracts'),
    'Symfony\\Component\\String\\' => array($vendorDir . '/symfony/string'),
    'Symfony\\Component\\Finder\\' => array($vendorDir . '/symfony/finder'),
    'Symfony\\Component\\Console\\' => array($vendorDir . '/symfony/console'),
    'Psr\\Container\\' => array($vendorDir . '/psr/container/src'),
    'MZ_Mobilize_America\\' => array($baseDir . '/src'),
    'League\\Flysystem\\' => array($vendorDir . '/league/flysystem/src'),
    'Composer\\Installers\\' => array($vendorDir . '/composer/installers/src/Composer/Installers'),
    'CoenJacobs\\Mozart\\' => array($vendorDir . '/coenjacobs/mozart/src'),
);

```

If not, please try running `composer dump`.

**Comment by MikeiLL** [comment]
I may have had some wonky namespace declarations going on in the Shortcode directory. Please pull in latest changes and try again. If not, perhaps look at above questions.

**Comment by jbradach** [comment]
Yes, I have `public/wp-content/plugins/mobilize-america/vendor/composer/autoload_psr4.php` with the same content.

I tried pulling in the changes but had the same result. Also tried tried a completely fresh `git clone`.

Error is the same.

**Comment by MikeiLL** [comment]
> Error is the same.

Oh yeah. Composer sure is making my life easier! Ha ha. I'll poke around at it ASAP. Thanks for your patience.

**Comment by MikeiLL** [comment]
I am able to "reproduce" the error if I remove the file at `src/Shortcode/Shortcode_Script_Loader.php`:

```
Fatal error: Uncaught Error: Class 'MZ_Mobilize_America\ShortCode\ShortCode_Script_Loader' not found in...
```

So maybe that's a start, but I doubt it.

I notice that in your server, it's using the ClassLoader located at `phar:///usr/local/bin/wp/vendor/composer/ClassLoader.php`, while mine is loading it from within the plugin codebase:

```
/var/www/html/wp-content/plugins/mobilize-america/vendor/composer/ClassLoader.php
```

Wondering if that might have something to do with the issue.

**Comment by MikeiLL** [comment]
I added a `/` to the end of the directory in the `psr4` configuration in `composer.json`.

Please try again. You may need to run `composer update`, `composer install` and/or `composer dump` again after pulling in the change.

Additionally, I, for the first time, ran `composer diagnose` and `composer self-update` on my local install, which seems to have been worth doing. At least I fixed some small config errors, I hadn't noticed.

**Comment by jbradach** [comment]
Activating the plugin through the admin page rather than WP-CLI command uses the same ClassLoader as you but still the not found error.

```php
Fatal error: Uncaught Error: Class 'MZ_Mobilize_America\ShortCode\ShortCode_Script_Loader' not found in /www/mobilize_211/public/wp-content/plugins/mobilize-america/src/Display/Display.php:9
Stack trace:
#0 /www/mobilize_211/public/wp-content/plugins/mobilize-america/vendor/composer/ClassLoader.php(444): include()
#1 /www/mobilize_211/public/wp-content/plugins/mobilize-america/vendor/composer/ClassLoader.php(322): Composer\Autoload\includeFile('/www/mobilize_2...')
#2 [internal function]: Composer\Autoload\ClassLoader->loadClass('MZ_Mobilize_Ame...')
#3 /www/mobilize_211/public/wp-content/plugins/mobilize-america/src/Core/Plugin_Core.php(142): spl_autoload_call('MZ_Mobilize_Ame...')
#4 /www/mobilize_211/public/wp-content/plugins/mobilize-america/src/Core/Plugin_Core.php(70): MZ_Mobilize_America\Core\Plugin_Core->define_public_hooks()
#5 /www/mobilize_211/public/wp-content/plugins/mobilize-america/mobilize-america.php(100): MZ_Mobilize_America\Core\Plugin_Core->__construct()
#6 /www/mobilize_211/public/wp-content/plugins/m in /www/mobilize_211/public/wp-content/plugins/mobilize-america/src/Display/Display.php on line 9
```

**Comment by MikeiLL** [comment]
And that's having pulled in my latest change?

Would you mind trying these two changes to confirm that psr4 autoload is getting the correct paths:

`autoload_real.php` around line 29:

`        $useStaticLoader = false; //PHP_VERSION_ID >= 50600 && !defined('HHVM_VERSION') && (!function_exists('zend_loader_file_encoded') || !zend_loader_file_encoded());
`

`autoload_psr4` around the top:

```
$vendorDir = dirname(dirname(__FILE__));
$baseDir = dirname($vendorDir);
echo "<hr>Some information</h1><pre>";
print_r($vendorDir . PHP_EOL);

print_r($baseDir . PHP_EOL);
print_r(array($baseDir . '/src'));
echo "</pre>";
die();
```

**Comment by jbradach** [comment]
Yes, I'm up to date.

```bash
mobilize@RZm-mobilize:~/public/wp-content/plugins/mobilize-america$ git pull
Already up to date.
```
After making the changes you indicated and trying to activate the plugin through the WordPress admin dashboard:
```
Plugin could not be activated because it triggered a fatal error.
Some information
/www/mobilize_275/public/wp-content/plugins/mobilize-america/vendor
/www/mobilize_275/public/wp-content/plugins/mobilize-america
Array
(
    [0] => /www/mobilize_275/public/wp-content/plugins/mobilize-america/src
)
```

**Comment by jbradach** [comment]
Happy to give you access to the site I'm using for this if it would help. It's a standalone site deployed explicitly for testing this.

**Comment by MikeiLL** [comment]
Cool. Those are the same paths I'm seeing in my psr4 script. I had killed to script after printing them to the screen.

Yeah if I can get on your server that would be great. Or at least might help. : ) 

I'll email you.

**Comment by MikeiLL** [comment]
The problem was that I was inconsistently using CamelCase (`ShortCode` vs `Shortcode`), which was working in my environment, but not in yours, which is apparently caseSensitive. Closed by [c09efaeb](https://github.com/MikeiLL/mobilize-america/commit/c09efaeb11c857604b271ac42c9f9ff602aef9d3)
